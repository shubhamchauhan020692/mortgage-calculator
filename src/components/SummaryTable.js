import React, { useState } from "react";
import { summaryCategories } from "../constants/SummaryCategories";

export const SummaryTable = () => {
  const data = [
    {
      category: [summaryCategories.NUMBER_OF_PAYMENTS],
      term: "36",
      amortization: "49",
    },
    {
      category: [summaryCategories.MORTGAGE_PAYMENT],
      term: "224",
      amortization: "429",
    },
    {
      category: [summaryCategories.PRINCIPAL_PAYMENTS],
      term: "55",
      amortization: "33",
    },
    {
      category: [summaryCategories.INTEREST_PAYMENTS],
      term: "36",
      amortization: "49",
    },
    {
      category: [summaryCategories.TOTAL_COST],
      term: "36",
      amortization: "49",
    },
  ];

  return (
    <table className="table table-striped">
      <thead className="thead-dark">
        <tr>
          <th scope="col">Category</th>
          <th scope="col">Term</th>
          <th scope="col">Amortization</th>
        </tr>
      </thead>
      <tbody>
        {data.map((item, index) => (
          <tr key={index}>
            <th scope="row">{item.category}</th>
            <td>{item.term}</td>
            <td>{item.amortization}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
