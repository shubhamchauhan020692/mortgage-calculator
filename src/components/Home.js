import React, { useState } from "react";
import { Charts } from "./Charts";
import { SummaryTable } from "./SummaryTable";
import { validateInputs } from "../utilities/validate";

import { calculateMortage } from "../utilities/calculations";
import { MortgageForm } from "./MortgageForm";
import { Header } from "./Header";

export const Home = () => {
  const [mortgageAmt, setMortgageAmt] = useState(50000);
  const [rate, setRate] = useState(5);
  const [amortizationPeriod, setAmortizationPeriod] = useState(25);
  const [paymentFrequency, setPaymentFrequency] = useState(12);
  const [term, setTerm] = useState(5);
  const [summaryTableData, setSummaryTableData] = useState(5);
  const [graphData, setGraphData] = useState(null);
  const [barData, setBarData] = useState({});
  const [errors, setErrors] = useState({
    mortgageAmt: "",
    rate: "",
    amortizationPeriod: "",
    paymentFrequency: "",
    term: "",
  });

  const handleInputChange = (e) => {
    setErrors({
      ...errors,
      [e.target.name]: "",
    });
    switch (e.target.name) {
      case "mortgageAmt": {
        setMortgageAmt(e.target.value);
        break;
      }
      case "rate": {
        setRate(e.target.value);
        break;
      }
      case "amortizationPeriod": {
        setAmortizationPeriod(e.target.value);
        break;
      }
      case "paymentFrequency": {
        setPaymentFrequency(e.target.value);
        break;
      }
      case "term": {
        setTerm(e.target.value);
        break;
      }
      default:
        return;
    }
  };

  const handleInputBlur = () => {
    const formErrors = validateInputs(
      mortgageAmt,
      rate,
      amortizationPeriod,
      term
    );
    formErrors &&
      setErrors({
        ...errors,
        ...formErrors,
      });

    /** TODO : If no error only then display */
    const { summaryTableData, graphData, barData } = calculateMortage(
      mortgageAmt,
      rate,
      amortizationPeriod,
      paymentFrequency,
      term
    );
    setSummaryTableData(summaryTableData);
    setGraphData(graphData);
    setBarData(barData);
  };

  console.log("errors", errors);
  return (
    <div className="main-container">
      <MortgageForm
        mortgageAmt={mortgageAmt}
        rate={rate}
        amortizationPeriod={amortizationPeriod}
        paymentFrequency={paymentFrequency}
        term={term}
        handleInputChange={handleInputChange}
        handleInputBlur={handleInputBlur}
        errors={errors}
      />
      <SummaryTable summaryTableData={summaryTableData} />
      {graphData && <Charts graphData={graphData} barData={barData} />}
    </div>
  );
};
