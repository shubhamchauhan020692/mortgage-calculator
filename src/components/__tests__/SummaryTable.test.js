import React from 'react';
import { fireEvent, cleanup, render } from '@testing-library/react';
import { SummaryTable } from '../SummaryTable';
import { mockSummaryTableData } from '../__mocks_/summaryTableData';
import { Home } from '../Home';

describe('Summary Table', () => {
  afterEach(cleanup);

  it('renders correctly', () => {
    const { asFragment } = render(<SummaryTable summaryTableData={mockSummaryTableData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Summary table correct data should be rendered', () => {
    const { queryByTestId, getAllByText, getByText } = render(<Home />);
    const mortgageAmt = queryByTestId('mortgageAmt')
    const rateInput = queryByTestId('rate')
    const amortizatoinPeriodInput = queryByTestId('amortizationPeriod')
    const paymentFrequencyInput = queryByTestId('paymentFrequency')
    const termInput = queryByTestId('term')
    fireEvent.change(mortgageAmt, { target: { value: '1000000' } });
    fireEvent.change(rateInput, { target: { value: '5' } });
    fireEvent.change(amortizatoinPeriodInput, { target: { value: '25' } });
    fireEvent.change(paymentFrequencyInput, { target: { value: '12' } });
    fireEvent.change(termInput, { target: { value: '5' } });
    fireEvent.blur(mortgageAmt);
    // console.log('getByTestId-000', getByTestId('term-0'))
    expect(getAllByText(/319.73/).length).toEqual(2)
    expect(getByText(/10019.40/)).toBeTruthy();
    expect(getByText(/28348.20/)).toBeTruthy();
    expect(getByText(/38367.78/)).toBeTruthy();
  });
});