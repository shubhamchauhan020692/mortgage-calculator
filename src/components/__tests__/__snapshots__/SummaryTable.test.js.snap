// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mortgage renders correctly 1`] = `
<DocumentFragment>
  <table
    class="table table-striped"
    data-testid="summary-table"
    id="summary-table"
  >
    <thead
      class="thead-dark"
    >
      <tr>
        <th
          scope="col"
        >
          Category
        </th>
        <th
          scope="col"
        >
          Term
        </th>
        <th
          scope="col"
        >
          Amortization
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Number of Payments
        </th>
        <td>
          120
        </td>
        <td>
          600
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Mortgage Payment
        </th>
        <td>
          319.73
        </td>
        <td>
          319.73
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Principal Payments,
        </th>
        <td>
          10019.40
        </td>
        <td>
          100000.00
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Interest Payments
        </th>
        <td>
          28348.20
        </td>
        <td>
          91838.89
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Total Cost
        </th>
        <td>
          38367.78
        </td>
        <td>
          191838.89
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
